### PR TITLE
Make MPIR_Get_Count_impl count safe, Fixes issue #2320

### PR DESCRIPTION
**Pull request by akhillanger**
Fix for Issue #2320. Explanation is there in the issue.
Tagging @jeffhammond for review.

### CODE REVIEW

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpi/coll/allgather/allgather_intra_recursive_doubling.c [hunk 1]
```diff
@@ -297,8 +298,12 @@ int MPIR_Allgather_intra_recursive_doubling(const void *sendbuf,
                     MPIR_ERR_SET(mpi_errno, *errflag, "**fail");
                     MPIR_ERR_ADD(mpi_errno_ret, mpi_errno);
                     last_recv_cnt = 0;
-                } else
-                    MPIR_Get_count_impl(&status, MPI_BYTE, &last_recv_cnt);
+                } else {
+                    byte_count = MPIR_STATUS_GET_COUNT(status);
+                    mpi_errno = MPIR_Get_elements_x_impl(&byte_count, MPI_BYTE, &last_recv_cnt);
```
Comment (raffenet) [inline src/mpi/coll/allgather/allgather_intra_recursive_doubling.c]: Is this really necessary when we already have the number of bytes in `byte_count`?  Seems redundant, no? IMO, a large-count safe `MPIR_Get_count_impl(const MPI_Status * status, MPI_Datatype datatype, MPI_Count *count)` would be more useful internally. `MPI_Get_count` would be responsible for truncating the value when > `INT_MAX`.

Comment (raffenet) [reply src/mpi/coll/allgather/allgather_intra_recursive_doubling.c]: Hmm, I misunderstood the usage of `MPIR_Get_elements_x_impl`. Nevertheless, my comment about a better `MPIR_Get_count_impl` remains.

Comment (jeffhammond) [reply src/mpi/coll/allgather/allgather_intra_recursive_doubling.c]: Yeah, I tend to agree.  We can change all the MPIR_Get_count_impl cases to the count-safe version rather easily.

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/ucx

**Comment by akhillanger** [comment]
I have applied the suggested solution - made MPIR_Get_count_impl count safe. Please review.

**Comment by akhillanger** [comment]
All tests passed.

**Comment by akhillanger** [comment]
Ok. I guess you are saying so because `MPIR_Get_count_impl` gets used mostly in `coll` directory and we have only `MPI_Aint` used in that directory. I used `MPI_Count` to be consistent with `MPIR_Get_elements_x_impl`

**Comment by akhillanger** [comment]
@raffenet I have changed MPI_Count to MPI_Aint.
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by raffenet** [comment]
@akhillanger thanks. I am fixing conflicts locally and will push for testing once I verify my changes.

**Comment by raffenet** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Pull request merged**